### PR TITLE
Assign a data path and serial number to each test in discover

### PR DESCRIPTION
**Pull request by happz**
To support both multihost scenarios and/or running a test multiple times, tmt needs to distinguish test data directories of different test instances. To allow this, each test instance is given a "serial number" which we can use in a data directory name.

Note: there can be more elaborate ways for picking the serial numbers, the patch starts with something really trivial yet functioning.

Another note: the patch does not isolate tests on a phase level, a single test, executed on different guests, may still land in the same data directory. Different patch will solve this issue.

Part of the effort in https://github.com/teemtee/tmt/issues/1614 and https://github.com/teemtee/tmt/pull/1790.

### CODE REVIEW

**Comment by lukaszachy** [comment]
Hm, when the test name is `/` the created workdir name is `-1`. That looks strange.

**Comment by happz** [comment]
> Hm, when the test name is `/` the created workdir name is `-1`. That looks strange. `/var/tmp/tmt/run-008/plans/default/execute/data/-1/output.txt`

Yeah, I'm adding some tests, I'll add this case as well. No idea how it happens though, but I'll look into it.

File: tmt/base.py [hunk 1]
```diff
@@ -1177,6 +1180,22 @@ def _expand_node_data(self, data: T, fmf_context: Dict[str, str]) -> T:
                 data[i] = self._expand_node_data(item, fmf_context)
         return data
 
+    # TODO: better, more elaborete ways of assigning serial numbers to tests
```
Comment (thrix) [inline tmt/base.py:1183]: typo: `elaborete`

File: tmt/base.py [hunk 2]
```diff
@@ -702,6 +703,8 @@ class Test(Core, tmt.export.Exportable['Test']):
     duration: str = DEFAULT_TEST_DURATION_L1
     result: str = 'respect'
 
+    serial_number: int = 0
```
Comment (thrix) [inline tmt/base.py:706]: I guess the docs need also an update?

https://tmt.readthedocs.io/en/stable/spec/plans.html#discover

File: tmt/result.py [hunk 3]
```diff
@@ -141,6 +142,7 @@ def from_test(
 
         _result = Result(
             name=test.name,
+            serial_number=test.serial_number,
```
Comment (thrix) [inline tmt/result.py:145]: @happz will `results.yaml` also change?

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/result.py (hunk 3 above)
Comment (happz) [reply tmt/result.py:145]: Yes, definitely. But, I have a problem with how `results.yaml` and `tests.yaml` are documented, and I plan to do something about it, including revamping the docs and adding JSON schemas for both. So, `results.yaml` does change, and `serial_number` starts appearing in results, but it's undocumented like some other fields are. If you won't mind, I'd leave this to a follow-up patch addressing documentation of both of these files in a more generous way.

File: tmt/base.py (hunk 1 above)
Comment (happz) [reply tmt/base.py]: Fixed in 495b7bf5

File: tmt/base.py (hunk 2 above)
Comment (happz) [reply tmt/base.py:706]: Fixed in 495b7bf5

**Review by lukaszachy — COMMENTED** [review]
LGTM to the code, still running tests/full in TJ#7580157

**Comment by lukaszachy** [comment]
Fails in tests/full:
/tests/discover/order
/tests/execute/reboot/reuse_provision
/tests/execute/filesubmit
/tests/execute/metadata
/tests/execute/upgrade/full

**Comment by happz** [comment]
Tests are green \o/

So, looking for reviews & merge, then:

* rebase https://github.com/teemtee/tmt/pull/1877 & fix tests
* submit a PR updating documentation of `tests.yaml` and `results.yaml`, to have it consistent before 1.22 gets out

File: tmt/result.py (hunk 3 above)
Comment (thrix) [reply tmt/result.py]: makes sense, ty!

**Pull request merged**